test/regress_ws.c

```diff
@@ -0,0 +1,362 @@
+#define ntohll(x) htonll(x)
+
+
+static struct event_base *exit_base;
+
+static void
+on_ws_msg_cb(struct evws_connection *evws, char *data, size_t len, void *arg)
+{
+	ev_uintptr_t val = (ev_uintptr_t)arg;
+
+	if (val != 0xDEADBEEF) {
+		fprintf(stdout, "FAILED on_complete_cb argument\n");
+		exit(1);
+	}
+
+	if (!strcmp(data, "Send echo")) {
+		const char *reply = "Reply echo";
+
+		evws_send(evws, reply, strlen(reply));
+		test_ok++;
+	} else if (!strcmp(data, "Client: hello")) {
+		test_ok++;
+	} else if (!strcmp(data, "Close")) {
+		evws_close(evws, 0);
+		test_ok++;
+	} else {
+		/* unexpected test message */
+		event_base_loopexit(arg, NULL);
+	}
+}
+
+static void
+on_ws_close_cb(struct evws_connection *evws, void *arg)
+{
+	ev_uintptr_t val = (ev_uintptr_t)arg;
+
+	if (val != 0xDEADBEEF) {
+		fprintf(stdout, "FAILED on_complete_cb argument\n");
+		exit(1);
+	}
+	test_ok++;
+}
+
+void
+http_on_ws_cb(struct evhttp_request *req, void *arg)
+{
+	struct evws_connection *evws;
+	const char *hello = "Server: hello";
+
+	evws = evws_new_session(req, on_ws_msg_cb, (void *)0xDEADBEEF);
+	if (!evws)
+		return;
+	test_ok++;
+
+	evws_connection_set_closecb(evws, on_ws_close_cb, (void *)0xDEADBEEF);
+	evws_send(evws, hello, strlen(hello));
+}
+
+static void
+http_ws_errorcb(struct bufferevent *bev, short what, void *arg)
+{
+	/** For ssl */
+	if (what & BEV_EVENT_CONNECTED)
+		return;
+	test_ok++;
+	event_base_loopexit(arg, NULL);
+}
+
+static char *
+receive_ws_msg(struct evbuffer *buf, size_t *out_len, unsigned *options)
+{
+	unsigned char *data;
+	int fin, opcode, mask;
+	uint64_t payload_len;
+	size_t header_len;
+	const unsigned char *mask_key;
+	char *out_buf = NULL;
+	size_t data_len = evbuffer_get_length(buf);
+
+	data = evbuffer_pullup(buf, data_len);
+
+	fin = !!(*data & 0x80);
+	opcode = *data & 0x0F;
+	mask = !!(*(data + 1) & 0x80);
+	payload_len = *(data + 1) & 0x7F;
+
+	header_len = 2 + (mask ? 4 : 0);
+
+	if (payload_len < 126) {
+		if (header_len > data_len)
+			return NULL;
+
+	} else if (payload_len == 126) {
+		header_len += 2;
+		if (header_len > data_len)
+			return NULL;
+
+		payload_len = ntohs(*(uint16_t *)(data + 2));
+
+	} else if (payload_len == 127) {
+		header_len += 8;
+		if (header_len > data_len)
+			return NULL;
+
+		payload_len = ntohll(*(uint64_t *)(data + 2));
+	}
+
+	if (header_len + payload_len > data_len)
+		return NULL;
+
+	mask_key = data + header_len - 4;
+	for (size_t i = 0; mask && i < payload_len; i++)
+		data[header_len + i] ^= mask_key[i % 4];
+
+	*out_len = payload_len;
+
+	/* text */
+	if (opcode == 0x01) {
+		out_buf = calloc(payload_len + 1, 1);
+	} else { /* binary */
+		out_buf = malloc(payload_len);
+	}
+	memcpy(out_buf, (const char *)data + header_len, payload_len);
+
+	if (!fin) {
+		*options = 1;
+	}
+
+	evbuffer_drain(buf, header_len + payload_len);
+	return out_buf;
+}
+
+static void
+send_ws_msg(struct evbuffer *buf, const char *msg)
```

Comment:
Maybe add some internal helper and use it here instead of doing the same encoding again?